app/models/classification_pipeline.rb

```diff
@@ -92,19 +93,29 @@ def reduce(workflow_id, subject_id, user_id)
     raise
   end
 
-  def check_rules(workflow_id, subject_id)
-    return unless rules.present?
+  def check_rules(workflow_id, subject_id, user_id)
```

Comment:
Hardly anything is shared between the subject and user part of this method. I'd suggest having it call two `check_subject_rules` and `check_user_rules`. Those two can then have the `return unless` guard clause, which will also mean we don't load subject reductions from the DB if there aren't any subject rules (but there are user rules).

Reply:
this does a lot for code clarity. i'm committing this change and then i'll merge.